### PR TITLE
[WIP] Additional features/customizations

### PR DESCRIPTION
**Pull request by thechrisnixon**
@Sharlaan started adding some features/customizations to the library as I've been using the library in a prod app. Let me know what you think of the changes. Still need to do some testing, add documentation, etc.

Allow optional/customizable select-all

Allow inner and outer styles on selections presenter

Allow custom footer height, autocomplete height & no match found height

Allow selections renderer to always be displayed and make the floating
label stay shrunken

### CODE REVIEW

**Review by Sharlaan — COMMENTED** [review]
Rename `alwaysDisplaySelectionsRenderer` with `floatingLabelFixed` to stay consistent with Material-UI naming.

File: src/SuperSelectField.js [hunk 1]
```diff
@@ -337,8 +350,18 @@ class SelectField extends Component {
...
 
   closeMenu = (reason) => {
-    const { onChange, name } = this.props
-    onChange(this.state.selectedItems, name)
+    const { onChange, name, allowSelectAll, selectAllItem } = this.props
+    let { selectedItems } = this.state
+    if (allowSelectAll) {
+      const selectAllItemIndex = selectedItems.findIndex(item => item.value === selectAllItem.value);
+      if (selectAllItemIndex !== -1) {
+        selectedItems = [
```
Comment (Sharlaan) [inline src/SuperSelectField.js]: You are mutating the state here, best practice is to first make a copy then play with it.

Also i don't understand what are you trying to achieve with these consecutive slices ?
Exclude the `selectAllItem` ? if yes, just use `.filter()`

File: src/SuperSelectField.js [hunk 2]
```diff
@@ -337,8 +350,18 @@ class SelectField extends Component {
   }
 
   closeMenu = (reason) => {
-    const { onChange, name } = this.props
-    onChange(this.state.selectedItems, name)
+    const { onChange, name, allowSelectAll, selectAllItem } = this.props
+    let { selectedItems } = this.state
```
Comment (Sharlaan) [inline src/SuperSelectField.js]: Should use `const` to make sure to never mutate the state.

File: src/SuperSelectField.js [hunk 3]
```diff
@@ -406,10 +429,35 @@ class SelectField extends Component {
     event.preventDefault()
     const { selectedItems } = this.state
     if (this.props.multiple) {
-      const selectedItemExists = selectedItems.some(obj => areEqual(obj.value, selectedItem.value))
-      const updatedValues = selectedItemExists
-        ? selectedItems.filter(obj => !areEqual(obj.value, selectedItem.value))
-        : selectedItems.concat(selectedItem)
+      const { allowSelectAll, selectAllItem } = this.props;
+      let updatedValues;
+      if (allowSelectAll && selectedItem.value === selectAllItem.value) {
```
Comment (Sharlaan) [inline src/SuperSelectField.js:431]: What is `selectAllItem` ? Can you document it in Readme.md please ? especially its value property.

Is it a bool ? if not, can you give a usecase please.

Comment (thechrisnixon) [reply src/SuperSelectField.js:431]: It is an object of the format `{ value: 'all-superSelectField', label: 'All' }`. It is in the PropTypes for the select field. I figured it _could_ be possible someone might already have an existing value that is `all`, so wanted to give them the ability to change it if necessary.

Is there a better name for it?

File: src/SuperSelectField.js (hunk 1 above)
Comment (thechrisnixon) [reply src/SuperSelectField.js]: I'll fix this soon, along with using `const` (comment below)

**Comment by thechrisnixon** [comment]
@Sharlaan I rebased my branch off of master, would you please do that for your branch too (if you're still working off of it)? Otherwise I can just re-point to `master`

**Comment by Sharlaan** [comment]
i forgot to tell you this:
would it be possible to recreate this PR with your own branch named, for exemple:
`dasnixon-selectall`
Also ideally should make 1 PR per feature, not 1 PR with 4-5 different packed features. Would be much easier to debug/accept/merge.

This branch `sharlaan-footer-renderer` is about some other feature totally offtopic

**Comment by thechrisnixon** [comment]
@Sharlaan will do, thanks for the feedback.

**Pull request closed without merging**